axum-extra/src/multipart_builder.rs

```diff
@@ -0,0 +1,337 @@
+    pub(super) fn serialize(&self) -> Vec<u8> {
+        // A part is serialized in this general format:
+        // // the filename is optional
+        // Content-Disposition: form-data; name="FIELD_NAME"; filename="FILENAME"\r\n
+        // // the mime type (not strictly required by the spec, but always sent here)
+        // Content-Type: mime/type\r\n
+        // // if the part does not conform to the rest of the request's encoding,
+        // // this is specified
+        // Content-Transfer-Encoding: "ENCODING"\r\n
+        // // a blank line, then the contents of the file start
+        // \r\n
+        // CONTENTS\r\n
+
+        // Format what we can as a string, then handle the rest at a byte level
+        let mut serialized_part = format!("Content-Disposition: form-data; name=\"{}\"", self.name);
+        // specify a filename if one was set
+        if let Some(filename) = &self.filename {
+            serialized_part += &format!("; filename=\"{}\"", filename);
+        }
+        serialized_part += "\r\n";
+        // specify the MIME type
+        serialized_part += &format!("Content-Type: {}\r\n", self.mime_type);
+        // if an encoding was set, add that
+        // determine what encoding to label the body of the field with
+        let encoding: Option<&str> = match self.encoding {
+            TransferEncoding::Default => None,
+            TransferEncoding::Binary => Some("binary"),
+        };
+        if let Some(encoding) = encoding {
+            serialized_part += &format!("Content-Transfer-Encoding: {}\r\n", encoding);
+        }
+        serialized_part += "\r\n";
+        let mut part_bytes = serialized_part.as_bytes().to_vec();
+        part_bytes.extend_from_slice(&self.contents);
+        part_bytes.extend_from_slice(b"\r\n");
+
+        part_bytes
+    }
+}
+
+/// A boundary is defined as a user defined (arbitrary) value that does not occur in any of the data.
+/// Because the specification does not clearly define a methodology for generating boundaries, this implementation
+/// follow's Reqwest's, and generates a boundary in the format of `XXXXXXXX-XXXXXXXX-XXXXXXXX-XXXXXXXX` where `XXXXXXXX`
+/// is a hexadecimal representation of a pseudo randomly generated u64.
+fn generate_boundary() -> String {
+    let a = gen_boundary_portion();
+    let b = gen_boundary_portion();
+    let c = gen_boundary_portion();
+    let d = gen_boundary_portion();
+    format!("{a:016x}-{b:016x}-{c:016x}-{d:016x}")
+}
+
+/// Generate a single portion of the boundary by performing an xor
+/// shift on the current time
+fn gen_boundary_portion() -> u64 {
+    // *very* rudimentary xor-shift used
+    // the current time is used as a seed because it doesn't really matter what the data is, it should
+    // just be vaguely random
+    let mut x = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis() as u64;
```

Comment:
Is this also copied from reqwest?

So we take system time, do 3 xors and shifts and repeat. I would be very surprised if this took a millisecond, I would even expect all of the four "pseudorandom" numbers to be the same one most of the time. Then again it's not really that probable the data would contain the same random `u64` in hex four times next to each other separated by a hyphen. So it serves its purpose.

If you want to keep this, just use the last result as the next seed I guess. But I don't think that using `rand` (or `fastrand`) would hurt performance here much. This seems like a case of premature optimization.

Reply:
No, they use their own implementation of an xor shift
https://github.com/seanmonstar/reqwest/blob/7e7011a657e2f97d3b52dda7bcd8af4353c06b8a/src/util.rs#L27

This impl is definitely a bit jank.

I used this here because frankly I tried to use `fastrand` first, but couldn't figure out how to make it work without CI errors. I am not experienced at all working with `cargo.toml` at this scale. I'll dig into that some more and try to figure it out, thanks for the feedback

Reply:
I figured this out, the impl uses fastrand now.